swarm/src/connection/pool.rs

```diff
@@ -52,12 +52,41 @@ use std::{
 use void::Void;
 
 mod concurrent_dial;
+#[cfg(any(feature = "tokio", feature = "async-std"))]
+pub mod executor;
 mod task;
 
+enum ExecSwitch {
+    Executor(Box<dyn Executor + Send>),
```

Comment:
I only just realised that `Executor` still lives in `libp2p-core`, despite being only used in `libp2p-swarm`!

Can you move it to a top-level `executor` within `libp2p-swarm` please?

The various implementations can then just go next to it (but please make sure they remain crate-private).

Reply:
I can move them to libp2p-swarm, but I would like to propose an alternative. I don't quite see why we are using our own custom Executor trait. Why not use [`futures::task::Spawn`](https://docs.rs/futures/0.3.25/futures/task/trait.Spawn.html#tymethod.spawn_obj)? With that we could simplify here and also get rid of every responsibilities regarding ThreadPool.

Reply:
> I can move them to libp2p-swarm, but I would like to propose an alternative. I don't quite see why we are using our own custom Executor trait. Why not use [`futures::task::Spawn`](https://docs.rs/futures/0.3.25/futures/task/trait.Spawn.html#tymethod.spawn_obj)? With that we could simplify here and also get rid of every responsibility regarding ThreadPool.

Can we? You still need to poll the `FuturesUnordered` so you can't completely erase the type.

I don't see how it makes things simpler but feel free to have a try!

> Another advantage: With the status method we could check if we can actually spawn tasks.

And if we can't spawn it then what? I don't see a reasonable error-handling strategy. Also, the interface guarantees are pretty weak so I don't see what benefit we are getting from "the executor is likely to accept the future".

Reply:
My gut feeling is that `Spawn` is not what we want but I am happy to be convinced otherwise :)